### PR TITLE
stop subclass dive if you hit bad import

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Recently I was facing losing potentially broken packages (see: https://github.com/Lightning-AI/lightning/pull/17829) and did not find other way around then this one, but happy to learn what I should do otherwise :chipmunk: 

## Before submitting

- [ ] Did you read the [contributing guideline](https://github.com/omni-us/jsonargparse/blob/master/CONTRIBUTING.rst)?
- [ ] Did you update **the documentation**? (readme and public docstrings)
- [ ] Did you write **unit tests** such that there is 100% coverage on related code? (required for bug fixes and new features)
- [x] Did you verify that new and existing **tests pass locally**?
- [x] Did you make sure that all changes preserve **backward compatibility**?
- [ ] Did you update **the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

### CODE REVIEW

**Review by mauvilsa — CHANGES_REQUESTED** [review]
To me here seems to be the correct place to fix it. It is actually a bug, because the `--help` of a CLI would crash.

Please add a unit test for get_all_subclass_paths mocking that an import fails. Also please add an entry to the changelog.

File: jsonargparse/typehints.py [hunk 1]
```diff
@@ -918,7 +918,10 @@ def is_private(class_path):
         return "._" in class_path
 
     def add_subclasses(cl):
-        class_path = get_import_path(cl)
+        try:
+            class_path = get_import_path(cl)
+        except (ImportError, AttributeError):  # Attribute is added in case of dot notation imports
+            return
```
Comment (Borda) [inline jsonargparse/typehints.py:924]: @mauvilsa shall we also print the error here? :thinking:

**Comment by Borda** [comment]
> Please add a unit test for get_all_subclass_paths mocking that an import fails

not sure how this mock properly as this with mock it calls `__original__` yielding in mock not in the mocked ImportError

**Comment by mauvilsa** [comment]
> > Please add a unit test for get_all_subclass_paths mocking that an import fails
> 
> not sure how this mock properly as this with mock it calls `__original__` yielding in mock not in the mocked ImportError

Could be something like:

```python
from unittest.mock import patch
from jsonargparse.util import get_import_path

class ImportClass:
    pass

def test_get_all_subclass_paths_import_error():

    def mocked_get_import_path(cls):
        if cls is ImportClass:
            raise ImportError
        return get_import_path(cls)

    with patch("jsonargparse.typehints.get_import_path", mocked_get_import_path):
        subclass_paths = get_all_subclass_paths(ImportClass)
    assert subclass_paths == []
```

File: jsonargparse_tests/test_typehints.py [hunk 2]
```diff
@@ -18,19 +18,22 @@
...
 )
+from unittest import mock
 
 import pytest
 import yaml
 
 from jsonargparse import ActionConfigFile, ArgumentError, Namespace, lazy_instance
-from jsonargparse.typehints import ActionTypeHint, Literal, is_optional
+from jsonargparse.typehints import ActionTypeHint, Literal, is_optional, resolve_class_path_by_name, \
+    get_all_subclass_paths
```
[the remark this answers is not in the document]
Comment (Borda) [reply jsonargparse_tests/test_typehints.py:28]:
```suggestion
from jsonargparse.typehints import ActionTypeHint, Literal, is_optional, \
    get_all_subclass_paths
```

File: jsonargparse/typehints.py [hunk 3]
```diff
@@ -918,7 +918,11 @@ def is_private(class_path):
         return "._" in class_path
 
     def add_subclasses(cl):
-        class_path = get_import_path(cl)
+        try:
+            class_path = get_import_path(cl)
+        except (ImportError, AttributeError) as err:  # Attribute is added in case of dot notation imports
+            warning(f"Hit failing import with following trace: {err}")
```
Comment (mauvilsa) [inline jsonargparse/typehints.py:924]:
```suggestion
            warning(f"Hit failing import with following error: {err}")
```

`{err}` is not a trace. It is only the message of the exception. The warning would show the line number here, but not a complete trace.

Comment (Borda) [reply jsonargparse/typehints.py]: so use `traceback` or another tool?
btw, I tried just printing the err and it was quite fine...

Comment (mauvilsa) [reply jsonargparse/typehints.py]: I don't think there is much benefit of printing the traceback here. Just the error message is good enough.

**Review by mauvilsa — APPROVED** [review]

**Pull request merged**